### ISSUE TITLE
perf(qwen3): decode sampling step tail is ~200us flat for FlashInfer sampling, 104us argmax at bs=256

### ISSUE THREAD

**Issue by n-WN** [xiaguan/pegainfer#483]
Found by the #456 roofline rank pass (H100 80GB, cold L2, `decode_sampling` manifest op measuring `openinfer_sample::select_batch` end to end, D2H readback + sync included — the production step tail).

| variant | bs=1 | bs=8 | bs=32 | bs=256 | analytic min (bs=256) |
|---|---:|---:|---:|---:|---:|
| `sampling` (FlashInfer temp/top-k/top-p) | 205µs | 322µs | 395µs | **1105µs** | 23µs |
| `argmax` (two-stage indexed argmax) | 31µs | 32µs | 41µs | **104µs** | 23µs |

Observations:

- The FlashInfer sampling path has a ~200µs *flat* cost before batch scaling even starts — at 0.0–2% of the logits-read speed-of-light everywhere. At bs=256 it is 1.1ms per decode step; against a ~6–9ms TPOT that is >10% of the step for every non-greedy request mix.
- `argmax` is fine at small bs (launch+readback floor) but grows to 104µs / 22% SoL at bs=256.

Worth splitting how much of the flat 200µs is FlashInfer plan/launch overhead vs the multi-pass rejection sampler itself, and whether the step tail can overlap the next step's metadata sync. Repro:

```
cargo run --release -p openinfer-qwen3-4b --features kernel-report --bin qwen3_kernel_report -- \
  run --op decode_sampling --no-cupti --out reports/decode_sampling.json
```

Relates to #456.

**Comment by n-WN** [comment]
Correction to the numbers above, from the adversarial-review pass on #485: the original measurements used near-flat synthetic logits, and the FlashInfer rejection sampler's round count tracks distribution flatness — so the flat vocabulary overstated its cost by 35–55%. Re-measured on peaked logits (top-1 mass ~0.5, row-varying peaks, same H100 / cold L2 / 128 iters):

| variant | bs=1 | bs=8 | bs=32 | bs=256 | analytic min (bs=256) |
|---|---:|---:|---:|---:|---:|
| `sampling` | ~~205µs~~ **95µs** | ~~322~~ **145µs** | ~~395~~ **208µs** | ~~1105~~ **722µs** | 23µs |
| `argmax` (distribution-insensitive) | 30µs | 33µs | 41µs | 103µs | 23µs |

The finding stands with the corrected magnitudes: a ~95µs flat cost, rising to 722µs at bs=256 — ~8–12% of a decode step and ~30× the logits-read bound. The flat-vs-peaked delta itself is informative for the fix: roughly half the measured cost at large bs was rejection rounds, which peaked real-world distributions largely avoid; the remaining flat ~95µs (plan/launch/multi-pass overhead + D2H tail) is the stable target.
